src/ripple/app/misc/detail/WorkBase.h

```diff
@@ -190,32 +191,39 @@ WorkBase<Impl>::fail(error_code const& ec)
 
 template <class Impl>
 void
-WorkBase<Impl>::onResolve(error_code const& ec, resolver_type::iterator it)
+WorkBase<Impl>::onResolve(error_code const& ec, results_type results)
 {
     if (ec)
         return fail(ec);
 
-    resolver_type::iterator end;
-    std::vector<
-        std::reference_wrapper<resolver_type::iterator::value_type const>>
-        entries;
-    bool inList = false;
-    for (; it != end; ++it)
-    {
-        if (lastEndpoint_ == *it)
-            inList = true;
-        else
-            entries.emplace_back(std::ref(*it));
-    }
     // Use last endpoint if it is successfully connected
     // and is in the list, otherwise pick a random endpoint
-    // from the list. If there is only one endpoint and
-    // it is the last endpoint then use the last endpoint.
-    if ((!lastStatus_ || !inList) && entries.size() > 0)
-        lastEndpoint_ =
-            entries[entries.size() > 1 ? rand_int(entries.size() - 1) : 0]
-                .get()
-                .endpoint();
+    // from the list (excluding last endpoint). If there is
+    // only one endpoint and it is the last endpoint then
+    // use the last endpoint.
+    lastEndpoint_ = [&]() -> endpoint_type {
+        int foundIndex = 0;
+        auto const foundIt = std::find_if(
+            results.begin(), results.end(), [&](endpoint_type const& e) {
+                if (e == lastEndpoint_)
+                    return true;
+                foundIndex++;
+                return false;
+            });
+        if (foundIt != results.end() && lastStatus_)
+            return lastEndpoint_;
+        else if (results.size() == 1)
+            return *results.begin();
+        else if (foundIt == results.end())
+            return *std::next(results.begin(), rand_int(results.size() - 1));
+
+        // We got more than 1 endpoint and last endpoint is in the list.
+        // If rand is equal to last endpoint then we select either
+        // previous or next endpoint.
+        auto r = rand_int(results.size() - 1);
+        return *std::next(
+            results.begin(), r != foundIndex ? r : (r > 0 ? --r : ++r));
```

Comment:
I think this is a biased sample. The `foundIndex`'s alternate element has two chances of being selected, while every other element has only one chance of being selected. Consider make the `r` not include the last element. If `foundIndex` is selected, then select the last element instead. This way every element has the same probability of being selected.

Reply:
You are right. Changed.